### PR TITLE
Gradient aggregation

### PR DESCRIPTION
**Pull request by aaron276h**
Enables GPU based gradient aggregation.

### CODE REVIEW

**Comment by aaron276h** [comment]
Still need to clean up some of the dead code ...

File: tensorpack/graph_builder/distributed.py [hunk 1]
```diff
@@ -277,27 +303,97 @@ def build(self, get_grad_fn, get_opt_fn):
...
+            for k in range(len(self.grad_list)):
+                for idx, (grad, var) in enumerate(self.grad_list[k]):
+                    with tf.device(self.raw_devices[k]), tf.variable_scope("aggregation_variables", reuse=True):
+                        grad_aggregation_variable_name = "task%dtower%didx%d" % (self.task_index, k, idx)
+                        grad_aggregator = tf.get_variable(grad_aggregation_variable_name)
+                        aggregation_ops_list.append(grad_aggregator.assign_add(grad))
+                        self.gpu_shadow_vars[k][idx] = (self.gpu_shadow_vars[k][idx][0], var)
+            aggregation_ops = tf.group(*aggregation_ops_list)
+        else:
+            aggregation_ops = tf.group(*[])
```
Comment (ddn0) [inline tensorpack/graph_builder/distributed.py]: ? isn't this just `tf.group()`

File: tensorpack/graph_builder/distributed.py [hunk 2]
```diff
@@ -277,27 +303,97 @@ def build(self, get_grad_fn, get_opt_fn):
...
-            ps_var_grads = DistributedReplicatedBuilder._apply_shadow_vars(avg_grads)
-            var_update_ops = self._apply_gradients_and_copy(
-                get_opt_fn(), grad_list, ps_var_grads)
-            self._shadow_vars = [v for (__, v) in ps_var_grads]
-            self._shadow_model_vars = DistributedReplicatedBuilder._shadow_model_variables(self._shadow_vars)
+        DataParallelBuilder._check_grad_list(self.grad_list)
+
+        if self.aggregation_frequency > 1:
+            # Create the variables which will be used to store aggregated updates
+            for k in range(len(self.grad_list)):
```
Comment (ddn0) [inline tensorpack/graph_builder/distributed.py]: `for k in range(len(self.gpu_shadow_vars), len(self.grad_list))` ?

File: tensorpack/graph_builder/distributed.py [hunk 3]
```diff
@@ -277,27 +303,97 @@ def build(self, get_grad_fn, get_opt_fn):
...
+                if k >= len(self.gpu_shadow_vars):
+                    self.gpu_shadow_vars.append([])
+                    for idx, (grad, var) in enumerate(self.grad_list[k]):
+                        if idx >= len(self.gpu_shadow_vars[k]):
+                            with tf.device(self.raw_devices[k]), tf.variable_scope("aggregation_variables"):
+                                grad_aggregation_variable_name = "task%dtower%didx%d" % (self.task_index, k, idx)
```
Comment (ddn0) [inline tensorpack/graph_builder/distributed.py]: don't know if this is consistent with the rest of the code base or if we are Python 3.6+, but `f"task{self.task_index}tower{k}idx{idx}"` might be more readable.

**Review by ddn0 — COMMENTED** [review]
Haven't thought too deeply about the logic part; something that might be useful is a simple test case / unit test that confirms the local aggregation (beyond just eyeballing print statements).

**Comment by aaron276h** [comment]
@ddn0 Yeah I agree having a unit test for this would be really good. Need to spend some time thinking about how it would be possible to do this.

File: tensorpack/graph_builder/distributed.py [hunk 4]
```diff
@@ -201,6 +201,24 @@ def __init__(self, towers, server):
...
+        # This is going to be K x N x 2 data structure holding the per-GPU aggregated updates
+        # and vars for parameter updates.
+        self.gpu_shadow_vars = []
+
+        # Used by comm_op to know when it can begin reading aggregated values.
+        self.counter = tf.ConditionalAccumulator(tf.float32)
+
+    def _combine_shadow_vars(self):
+        assert len(self.grad_accumulators) == len(self.vars)
+        ps_var_grads = []
```
Comment (dzhu) [inline tensorpack/graph_builder/distributed.py]: `return [(acc.take_grad(self.aggregation_frequency), var) for acc, var in zip(self.grad_accumulators, self.vars)]`

File: tensorpack/graph_builder/distributed.py [hunk 5]
```diff
@@ -277,27 +295,99 @@ def build(self, get_grad_fn, get_opt_fn):
...
-            grad_list, colocation=False, devices=self.raw_devices)
-        with tf.device(self.param_server_device):
-            ps_var_grads = DistributedReplicatedBuilder._apply_shadow_vars(avg_grads)
-            var_update_ops = self._apply_gradients_and_copy(
-                get_opt_fn(), grad_list, ps_var_grads)
-            self._shadow_vars = [v for (__, v) in ps_var_grads]
-            self._shadow_model_vars = DistributedReplicatedBuilder._shadow_model_variables(self._shadow_vars)
+        DataParallelBuilder._check_grad_list(self.grad_list)
+
+        if self.aggregation_frequency > 1:
```
Comment (dzhu) [inline tensorpack/graph_builder/distributed.py:303]: Seems like this could just be
```python
aggregation_ops = []
if self.aggregation_frequency > 1:
    ...
with tf.control_dependencies(aggregation_ops):
    ...
```
unless `tf.group` has some extra semantics I can't see.

File: tensorpack/graph_builder/distributed.py [hunk 6]
```diff
@@ -277,27 +295,99 @@ def build(self, get_grad_fn, get_opt_fn):
...
+                        grad_aggregation_variable = tf.get_variable(
+                            grad_aggregation_variable_name, shape=grad.get_shape().as_list(),
+                            trainable=False, initializer=tf.zeros_initializer(),
+                            collections=[tf.GraphKeys.LOCAL_VARIABLES, "aggregating_collection"])
+                        self.gpu_shadow_vars[k].append((grad_aggregation_variable, None))
+            assert len(self.gpu_shadow_vars) == len(self.grad_list)
+
+            # Apply new gradients to variables.
+            aggregation_ops_list = []
+            for k in range(len(self.grad_list)):
```
Comment (dzhu) [inline tensorpack/graph_builder/distributed.py]:
```python
for k, (grad, device, gpu_vars) in enumerate(zip(self.grad_list, self.raw_devices, self.gpu_shadow_vars)):
    ...
```

File: tensorpack/graph_builder/distributed.py [hunk 7]
```diff
@@ -201,6 +201,24 @@ def __init__(self, towers, server):
         # Device for queues for managing synchronization between servers
         self.sync_queue_devices = ['/job:ps/task:%s/cpu:0' % i for i in range(self.num_ps)]
 
+        # How often are parameters synchronized
+        self.aggregation_frequency = aggregation_frequency
+        assert self.aggregation_frequency > 0
+
+        # This is going to be K x N x 2 data structure holding the per-GPU aggregated updates
```
Comment (dzhu) [inline tensorpack/graph_builder/distributed.py:208]: Say what K and N are and what the 2 signifies.

File: tensorpack/graph_builder/distributed.py [hunk 8]
```diff
@@ -277,27 +295,99 @@ def build(self, get_grad_fn, get_opt_fn):
...
-            ps_var_grads = DistributedReplicatedBuilder._apply_shadow_vars(avg_grads)
-            var_update_ops = self._apply_gradients_and_copy(
-                get_opt_fn(), grad_list, ps_var_grads)
-            self._shadow_vars = [v for (__, v) in ps_var_grads]
-            self._shadow_model_vars = DistributedReplicatedBuilder._shadow_model_variables(self._shadow_vars)
+        DataParallelBuilder._check_grad_list(self.grad_list)
+
+        if self.aggregation_frequency > 1:
+            # Create the variables which will be used to store aggregated updates.
+            for k in range(len(self.gpu_shadow_vars), len(self.grad_list)):
```
Comment (dzhu) [inline tensorpack/graph_builder/distributed.py]: If we only expect `gpu_shadow_vars` to be either empty or already the same length as `grad_list`, I would make that explicit:
```python
if not self.gpu_shadow_vars:
    ... # build self.gpu_shadow_vars
assert len(self.gpu_shadow_vars) == len(self.grad_list)
```
Also, can `gpu_shadow_vars` already be non-empty here? I guess if `build` is called multiple times; can that happen? Any other way?

File: tensorpack/graph_builder/distributed.py [hunk 9]
```diff
@@ -277,27 +295,99 @@ def build(self, get_grad_fn, get_opt_fn):
...
+                            tower_id=k, variable_id=idx)
+                        grad_aggregator = tf.get_variable(grad_aggregation_variable_name)
+                        aggregation_ops_list.append(grad_aggregator.assign_add(grad))
+                        self.gpu_shadow_vars[k][idx] = (self.gpu_shadow_vars[k][idx][0], var)
+            aggregation_ops = tf.group(*aggregation_ops_list)
+        else:
+            aggregation_ops = tf.group()
+
+        with tf.control_dependencies([aggregation_ops]):
+            train_op = self.counter.apply_grad(tf.constant([1], dtype=tf.float32), local_step=990000000)
```
Comment (dzhu) [inline tensorpack/graph_builder/distributed.py:335]: Comment that the step number is just a very large number so that it's always greater than the counter's step.

File: tensorpack/graph_builder/distributed.py [hunk 10]
```diff
@@ -308,7 +398,22 @@ def build(self, get_grad_fn, get_opt_fn):
...
+        variable.
+
+        Args:
+            tower_id: index of the GPU within the machine
+            variable_id: index of the variable
+
+        Returns:
+            Variable name.
+        """
+        grad_aggregation_variable_name = f"task{self.task_index}tower{tower_id}idx{variable_id}"
```
Comment (dzhu) [inline tensorpack/graph_builder/distributed.py]: `return f"..."`

File: tensorpack/train/base.py [hunk 11]
```diff
@@ -255,7 +258,7 @@ def initialize_hooks(self):
             session_creator=ReuseSessionCreator(self.sess), hooks=hooks)
 
     @call_only_once
-    def main_loop(self, steps_per_epoch, starting_epoch, max_epoch):
+    def main_loop(self, steps_per_epoch, starting_epoch, max_epoch, aggregation_frequency):
```
Comment (dzhu) [inline tensorpack/train/base.py:261]: This could avoid having to take the frequency again by looking for `self._builder.aggregation_frequency`, defaulting to 1 if it's not present. (Unless passing in a different value here is ever reasonable?)

File: tensorpack/graph_builder/distributed.py [hunk 12]
```diff
@@ -330,6 +435,7 @@ def _apply_gradients_and_copy(self, opt, raw_grad_list, ps_var_grads):
                 apply_gradient_op = opt.apply_gradients([(g, v)])
                 barrier = self._add_sync_queues_and_barrier(
                     'param_update_barrier_{}'.format(vid), [apply_gradient_op])
+
```
Comment (dzhu) [inline tensorpack/graph_builder/distributed.py]: As a general rule, I would say to make sure to remove all these extraneous lines and the other unnecessary changes in spacing, though I'm not going to press much in this situation.

File: tensorpack/graph_builder/distributed.py [hunk 13]
```diff
@@ -277,27 +295,99 @@ def build(self, get_grad_fn, get_opt_fn):
...
+            avg_grads = aggregate_grads(
+                aggregated_grads, colocation=False,
+                devices=self.raw_devices, average=False,
+                aggregation_frequency=self.aggregation_frequency)
+
+            with tf.device(self.param_server_device):
+                ps_var_grads = DistributedReplicatedBuilder._apply_shadow_vars(avg_grads)
+                var_update_ops = self._apply_gradients_and_copy(
+                    get_opt_fn(), self.grad_list, ps_var_grads)
+                self._shadow_vars = [v for (__, v) in ps_var_grads]
```
Comment (dzhu) [inline tensorpack/graph_builder/distributed.py]: `v for _, v`

**Review by dzhu — COMMENTED** [review]
Here are some general code style things; let's meet to discuss the actual logic here.

File: tensorpack/graph_builder/distributed.py [hunk 14]
```diff
@@ -277,27 +295,99 @@ def build(self, get_grad_fn, get_opt_fn):
...
+                    for idx, (grad, var) in enumerate(self.gpu_shadow_vars[k]):
+                        with tf.device(self.raw_devices[k]), tf.variable_scope("aggregation_variables", reuse=True):
+                            grad_aggregation_variable_name = self._get_grad_aggregator_variable_name(
+                                tower_id=k, variable_id=idx)
+                            grad_aggregator = tf.get_variable(grad_aggregation_variable_name)
+                            clear_op = grad_aggregator.assign(grad_aggregator.initial_value)
+                            clear_ops_list.append(clear_op)
+            clear_ops = tf.group(*clear_ops_list)
+        else:
+            with tf.control_dependencies([ready_to_communicate]):
```
Comment (dzhu) [inline tensorpack/graph_builder/distributed.py]: I don't think this `with` actually does anything, since no tensors/ops are being created inside. Just assigning stuff doesn't show up to TF.

**Comment by aaron276h** [comment]
Created DET-793 to make sure tests will be implemented in the future.

File: tensorpack/graph_builder/distributed.py (hunk 8 above)
Comment (aaron276h) [reply tensorpack/graph_builder/distributed.py]: Yeah it would only happen if build is called more than once. I am not sure under what scenario this would occur(and if our code would work), so going to just put an assert here checking that `self.gpu_shadow_vars` are empty before we create them.

File: tensorpack/train/base.py [hunk 15]
```diff
@@ -364,6 +365,11 @@ def __new__(cls, *args, **kwargs):
         else:
             return super(Trainer, cls).__new__(cls)
 
+    def _get_aggregation_frequency(self):
+        if hasattr(self._builder, 'aggregation_frequency'):
```
Comment (dzhu) [inline tensorpack/train/base.py:371]: `return getattr(self._builder, 'aggregation_frequency', 1)`

(Just to check, is there definitely always a `_builder`?)

File: tensorpack/graph_builder/distributed.py [hunk 16]
```diff
@@ -383,7 +380,7 @@ def build(self, get_grad_fn, get_opt_fn):
                 ps_var_grads = DistributedReplicatedBuilder._apply_shadow_vars(avg_grads)
                 var_update_ops = self._apply_gradients_and_copy(
                     get_opt_fn(), self.grad_list, ps_var_grads)
-                self._shadow_vars = [v for (__, v) in ps_var_grads]
+                self._shadow_vars = [v for __, v in ps_var_grads]
```
Comment (dzhu) [inline tensorpack/graph_builder/distributed.py]: Single underscore.

File: tensorpack/graph_builder/distributed.py (hunk 8 above)
Comment (dzhu) [reply tensorpack/graph_builder/distributed.py]: Okay, looks good. The loop can now be `for k, (gpu_grads, device) in enumerate(zip(self.grad_list, self.raw_devices))`.

File: tensorpack/graph_builder/distributed.py [hunk 17]
```diff
@@ -319,21 +316,21 @@ def build(self, get_grad_fn, get_opt_fn):
...
+                    with tf.device(device), tf.variable_scope("aggregation_variables", reuse=True):
                         grad_aggregation_variable_name = self._get_grad_aggregator_variable_name(
                             tower_id=k, variable_id=idx)
                         grad_aggregator = tf.get_variable(grad_aggregation_variable_name)
                         aggregation_ops_list.append(grad_aggregator.assign_add(grad))
                         self.gpu_shadow_vars[k][idx] = (self.gpu_shadow_vars[k][idx][0], var)
-            aggregation_ops = tf.group(*aggregation_ops_list)
-        else:
-            aggregation_ops = tf.group()
+        aggregation_ops = tf.group(*aggregation_ops_list)
```
Comment (dzhu) [inline tensorpack/graph_builder/distributed.py:329]: So is the `tf.group` necessary for some reason or can you pass the list directly to `control_dependencies`?

File: tensorpack/graph_builder/distributed.py [hunk 18]
```diff
@@ -206,19 +206,14 @@ def __init__(self, towers, server, aggregation_frequency):
         assert self.aggregation_frequency > 0
 
         # This is going to be K x N x 2 data structure holding the per-GPU aggregated updates
-        # and vars for parameter updates.
+        # and vars for parameter updates. K is the number of GPUs on the machine. N is the number
+        # of parameters, and there are 2 entries per parameter because each entry the gradient
```
Comment (dzhu) [inline tensorpack/graph_builder/distributed.py]: "each entry contains"

File: tensorpack/graph_builder/distributed.py [hunk 19]
```diff
@@ -277,27 +290,101 @@ def build(self, get_grad_fn, get_opt_fn):
...
+        DataParallelBuilder._check_grad_list(self.grad_list)
+
+        aggregation_ops_list = []
+        if self.aggregation_frequency > 1:
+            assert not self.gpu_shadow_vars
+            # Create the variables which will be used to store aggregated updates.
+            for k in range(len(self.gpu_shadow_vars), len(self.grad_list)):
+                self.gpu_shadow_vars.append([])
+                for idx, (grad, var) in enumerate(self.grad_list[k]):
+                    with tf.device(self.raw_devices[k]), tf.variable_scope("aggregation_variables"):
```
Comment (dzhu) [inline tensorpack/graph_builder/distributed.py]: Swap the `with` and the `for`.

File: tensorpack/graph_builder/distributed.py [hunk 20]
```diff
@@ -319,21 +316,21 @@ def build(self, get_grad_fn, get_opt_fn):
...
 
-            # Apply new gradients to variables.
-            aggregation_ops_list = []
-            for k in range(len(self.grad_list)):
-                for idx, (grad, var) in enumerate(self.grad_list[k]):
-                    with tf.device(self.raw_devices[k]), tf.variable_scope("aggregation_variables", reuse=True):
+            # Apply new gradients to aggregation variables.
+            for k, (grad_list, device) in enumerate(zip(self.grad_list, self.raw_devices)):
+                for idx, (grad, var) in enumerate(grad_list):
+                    with tf.device(device), tf.variable_scope("aggregation_variables", reuse=True):
```
Comment (dzhu) [inline tensorpack/graph_builder/distributed.py]: Swap `with` and `for`.

File: tensorpack/graph_builder/utils.py [hunk 21]
```diff
@@ -264,9 +266,9 @@ def aggregate_grads(all_grads,
 
     def aggregate(grads):
         if average:
-            return tf.multiply(tf.add_n(grads), 1.0 / nr_tower)
+            return tf.multiply(tf.add_n(grads), 1.0 / (nr_tower * aggregation_frequency))
         else:
-            return tf.add_n(grads)
+            return  tf.multiply(tf.add_n(grads), 1.0 / aggregation_frequency)
```
Comment (dzhu) [inline tensorpack/graph_builder/utils.py]: Extra space.

File: tensorpack/graph_builder/distributed.py [hunk 22]
```diff
@@ -277,27 +290,101 @@ def build(self, get_grad_fn, get_opt_fn):
...
+                            grad_aggregator = tf.get_variable(grad_aggregation_variable_name)
+                            tower_aggregated_grads.append((grad_aggregator.read_value(), var))
+                            aggregation_read_ops_list.append(tower_aggregated_grads[idx][0].op)
+                    aggregated_grads.append(tower_aggregated_grads)
+            aggregation_read_ops = tf.group(*aggregation_read_ops_list)
+
+            # Clear gradients.
+            clear_ops_list = []
+            with tf.control_dependencies([aggregation_read_ops]):
+                for k in range(len(self.gpu_shadow_vars)):
```
Comment (dzhu) [inline tensorpack/graph_builder/distributed.py]: Use `enumerate`.

File: tensorpack/graph_builder/distributed.py [hunk 23]
```diff
@@ -277,27 +290,101 @@ def build(self, get_grad_fn, get_opt_fn):
...
+
+        # Communication op begins here.
+        ready_to_communicate = self.counter.take_grad(self.aggregation_frequency)
+
+        if self.aggregation_frequency > 1:
+            # Read in latest variables values.
+            aggregated_grads = []
+            aggregation_read_ops_list = []
+            with tf.control_dependencies([ready_to_communicate]):
+                for k in range(len(self.gpu_shadow_vars)):
```
Comment (dzhu) [inline tensorpack/graph_builder/distributed.py]: Use `enumerate`.

File: tensorpack/graph_builder/distributed.py [hunk 24]
```diff
@@ -277,27 +290,101 @@ def build(self, get_grad_fn, get_opt_fn):
...
+                    tower_aggregated_grads = []
+                    for idx, (grad, var) in enumerate(self.gpu_shadow_vars[k]):
+                        with tf.device(self.raw_devices[k]), tf.variable_scope("aggregation_variables", reuse=True):
```
Comment (dzhu) [inline tensorpack/graph_builder/distributed.py]: Swap `with` and `for`.

**Review by dzhu — COMMENTED** [review]
Few more style things I want to see, but the interface is good, the logic seems right, and, well, if it works, it works.

File: tensorpack/graph_builder/distributed.py (hunk 17 above)
Comment (aaron276h) [reply tensorpack/graph_builder/distributed.py:329]: I am using a `tf.group()` here so that it matches the style of the tensorpack code, can drop since it's possible to just pass a list into `control_dependencies`

File: tensorpack/graph_builder/distributed.py (hunk 14 above)
Comment (aaron276h) [reply tensorpack/graph_builder/distributed.py]: Yeah I wasn't sure about how TF treats this, decided to leave it in just in case, but we can drop it if you think that's ok.

File: tensorpack/train/base.py (hunk 15 above)
Comment (aaron276h) [reply tensorpack/train/base.py:371]: I checked that `SyncMultiGPUTrainerReplicated` has a `_builder` and that is the only other trainer we currently support. But it does seem like they all have it except Horovod and BytePSTrainer. Can put in checks for this also.

Comment (dzhu) [reply tensorpack/train/base.py:371]: We may as well have a check for it just in case. So just `if not hasattr(self, '_builder'): return 1`.

File: tensorpack/graph_builder/distributed.py (hunk 17 above)
Comment (dzhu) [reply tensorpack/graph_builder/distributed.py:329]: I'll advocate for the simpler code as a general principle, but if it looks like other stuff here, then either way is fine.

File: tensorpack/graph_builder/distributed.py (hunk 14 above)
Comment (dzhu) [reply tensorpack/graph_builder/distributed.py]: Yeah, take it out.

File: tensorpack/graph_builder/distributed.py [hunk 25]
```diff
@@ -277,27 +290,101 @@ def build(self, get_grad_fn, get_opt_fn):
...
+
+        # Communication op begins here.
+        ready_to_communicate = self.counter.take_grad(self.aggregation_frequency)
+
+        if self.aggregation_frequency > 1:
+            # Read in latest variables values.
+            aggregated_grads = []
+            aggregation_read_ops_list = []
+            with tf.control_dependencies([ready_to_communicate]):
+                for k,_ in enumerate(self.gpu_shadow_vars):
```
Comment (dzhu) [inline tensorpack/graph_builder/distributed.py]: `for k, (vars, device) in enumerate(zip(self.gpu_shadow_vars, self.raw_devices))`

Then replace `self.gpu_shadow_vars[k]` with `vars` and `self.raw_devices[k]` with `device`.

This should define a variable instead of `_` and use it below instead of reindexing into `gpu_shadow_vars`.

File: tensorpack/graph_builder/distributed.py [hunk 26]
```diff
@@ -277,27 +290,101 @@ def build(self, get_grad_fn, get_opt_fn):
...
+                            grad_aggregator = tf.get_variable(grad_aggregation_variable_name)
+                            tower_aggregated_grads.append((grad_aggregator.read_value(), var))
+                            aggregation_read_ops_list.append(tower_aggregated_grads[idx][0].op)
+                    aggregated_grads.append(tower_aggregated_grads)
+            aggregation_read_ops = tf.group(*aggregation_read_ops_list)
+
+            # Clear gradients.
+            clear_ops_list = []
+            with tf.control_dependencies([aggregation_read_ops]):
+                for k,_ in enumerate(self.gpu_shadow_vars):
```
Comment (dzhu) [inline tensorpack/graph_builder/distributed.py]: Same as above.

File: tensorpack/graph_builder/distributed.py [hunk 27]
```diff
@@ -277,27 +290,101 @@ def build(self, get_grad_fn, get_opt_fn):
...
+
+            # Apply new gradients to aggregation variables.
+            for k, (grad_list, device) in enumerate(zip(self.grad_list, self.raw_devices)):
+                with tf.device(device), tf.variable_scope("aggregation_variables", reuse=True):
+                    for idx, (grad, var) in enumerate(grad_list):
+                        grad_aggregation_variable_name = self._get_grad_aggregator_variable_name(
+                            tower_id=k, variable_id=idx)
+                        grad_aggregator = tf.get_variable(grad_aggregation_variable_name)
+                        aggregation_ops_list.append(grad_aggregator.assign_add(grad))
+                        self.gpu_shadow_vars[k][idx] = (self.gpu_shadow_vars[k][idx][0], var)
```
Comment (dzhu) [inline tensorpack/graph_builder/distributed.py]: Include `self.gpu_shadow_vars` in the `enumerate`.

File: tensorpack/graph_builder/distributed.py [hunk 28]
```diff
@@ -277,27 +290,101 @@ def build(self, get_grad_fn, get_opt_fn):
...
+                self.gpu_shadow_vars.append([])
+                with tf.device(device), tf.variable_scope("aggregation_variables"):
+                    for idx, (grad, var) in enumerate(grad_list):
+                        grad_aggregation_variable_name = self._get_grad_aggregator_variable_name(
+                            tower_id=k, variable_id=idx)
+                        grad_aggregation_variable = tf.get_variable(
+                            grad_aggregation_variable_name, shape=grad.get_shape().as_list(),
+                            trainable=False, initializer=tf.zeros_initializer(),
+                            collections=[tf.GraphKeys.LOCAL_VARIABLES, "aggregating_collection"])
+                        self.gpu_shadow_vars[k].append((grad_aggregation_variable, None))
```
Comment (dzhu) [inline tensorpack/graph_builder/distributed.py]: `-1` instead of `k`.

**Review by dzhu — COMMENTED** [review]
I realize it's getting pretty late for you, so let me know if you want me to finish it up to get it over with before Yoav gets ready for testing.

**Comment by aaron276h** [comment]
I can do these changes, will push in a bit. 

Edit: Done.
